### DISCUSSION TITLE
Mixing Legacy and ForceTorque Power Distribution

### DISCUSSION THREAD

**Discussion by jafriztrillo** [bitcraze/discussions#1363] [Q&A]
I am in the process of implementing a controller that replaces each PID control block so the entire control structure has 6 controllers in it. I want to implement the controller in the z-direction first since it is decoupled from the rest of the quadcopter control architecture to try and isolate errors in the controller. The issue I am running into is the control design uses forces and I want to integrate it with the default PID controllers for x, y, and attitudes. 

My current idea is to use the default method from powerDistributionLegacy to calculate the attitudes with
```
  int16_t r = control->roll / 2.0f;
  int16_t p = control->pitch / 2.0f;
  int16_t y = control->yaw;
```
then calculate thrust by 
```
  const float thrustPart = 0.25f * control->thrustSi; // N (per rotor)
float motor_pwm_force = (-pwmToThrustB  + sqrtf(pwmToThrustB * pwmToThrustB + 4.0f * pwmToThrustA * motorForce)) / (2.0f * pwmToThrustA);
```
and adding each of the control efforts together since they should all be in the same unit scale. Something about this process seems wrong, but I can't quite figure out what it is or how to figure out a better method of combining these two. Any help on doing this combining better?

**Comment by gemenerik** [comment]
You're defining several variables. However, these variables (like r, p, y, and thrustPart) don't seem to be used in the rest of your code. Could you clarify what your main goal is with these variables? Specifically, are you trying to output forces, or are you looking to achieve something else with your controller? Additionally, could you explain why you're dividing the roll and pitch by two? It would be very helpful if you could rewrite your question to specify exactly what you're trying to do and how you want to combine your custom controller with the existing PID controllers.
